cpu/cpu_hotplug.py

```diff
@@ -0,0 +1,126 @@
+import random
+import time
+from avocado import Test
+from avocado import main
+from avocado.utils import process, cpu
+
+
+class cpuhotplug_test(Test):
+    """
+    To test hotplug within core in random manner.
+    """
+    def setUp(self):
+        """
+        Get the number of cores and threads per core
+        Set the SMT value to 4/8
+        """
+        command = "uname -p"
+        if 'ppc' not in process.system_output(command, ignore_status=True):
+            self.cancel("Processor is not ppc64")
+        self.nfail = 0
+        self.CORES = process.system_output("lscpu | grep 'Core(s) per socket:'"
+                                           "| awk '{print $4}'", shell=True)
+        self.SOCKETS = process.system_output("lscpu | grep 'Socket(s):'"
+                                             "| awk '{print $2}'", shell=True)
+        self.THREADS = process.system_output("lscpu | grep 'Thread(s) per core"
+                                             ":'| awk '{print $4}'",
+                                             shell=True)
+        self.T_CORES = int(self.CORES) * int(self.SOCKETS)
+        self.log.info(" Cores = %s and threads = %s "
+                      % (self.T_CORES, self.THREADS))
+        process.system_output("echo 8 > /proc/sys/kernel/printk", shell=True)
+        self.max_smt = 4
+        if cpu.get_cpu_arch().lower() == 'power8':
+            self.max_smt = 8
+        if cpu.get_cpu_arch().lower() == 'power6':
+            self.max_smt = 2
+        process.system_output("ppc64_cpu --smt=%s" % self.max_smt, shell=True)
+        self.path = "/sys/devices/system/cpu"
+
+    def test(self):
+        """
+        This script picks a random core and then offlines all its threads
+        in a random order and onlines all its threads in a random order.
+        """
+        for x in range(1, 100):
+            self.log.info("================= TEST %s ==================" % x)
+            core_list = []
+            core_list = self.random_gen_cores()
+            for core in core_list:
+                cpu_list = []
+                cpu_list = self.random_gen_cpu(core)
+                self.log.info("Offlining the threads : %s for "
+                              "the core : %s" % (cpu_list, core))
+                for cpu_num in cpu_list:
+                    self.offline_cpu(cpu_num)
+                cpu_list = self.random_gen_cpu(core)
+                self.log.info("Onlining the threads : %s for "
+                              "the core : %s" % (cpu_list, core))
+                for cpu_num in cpu_list:
+                    self.online_cpu(cpu_num)
+        if self.nfail > 0:
+            self.fail(" Unable to online/offline few cpus")
+
+    def random_gen_cores(self):
+        """
+        Generate random core list
+        """
+        nums = [x for x in range(0, self.T_CORES)]
+        random.shuffle(nums)
+        self.log.info(" Core list is %s" % nums)
+        return nums
+
```

Comment:
To generate a random number you can use 
from random import randint
core = randint(0, cores)

Reply:
I need list here, so I use the above logic which gives shuffled core list.
Ex : 
[stdout] [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23]
 Core list is [22, 18, 8, 11, 0, 12, 2, 1, 23, 7, 10, 17, 19, 6, 20, 5, 4, 13, 14, 15, 16, 3, 9, 21]